src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp

```diff
@@ -15,577 +16,1264 @@
+    }
 
-        void SetData(uint32 uiType, uint32 /*uiData*/) override
+    bool GrandChampionsDefeated()
+    {
+        for (int i = 0; i < 3; i++)
         {
-            switch (uiType)
+            if (Creature* pGrandChampion = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_GRAND_CHAMPION_1 + i)))
             {
-                case 1:
-                    AddWaypoint(0, 747.36f, 634.07f, 411.572f);
-                    AddWaypoint(1, 780.43f, 607.15f, 411.82f);
-                    AddWaypoint(2, 785.99f, 599.41f, 411.92f);
-                    AddWaypoint(3, 778.44f, 601.64f, 411.79f);
-                    uiWaypointPath = 1;
-                    break;
-                case 2:
-                    AddWaypoint(0, 747.35f, 634.07f, 411.57f);
-                    AddWaypoint(1, 768.72f, 581.01f, 411.92f);
-                    AddWaypoint(2, 763.55f, 590.52f, 411.71f);
-                    uiWaypointPath = 2;
-                    break;
-                case 3:
-                    AddWaypoint(0, 747.35f, 634.07f, 411.57f);
-                    AddWaypoint(1, 784.02f, 645.33f, 412.39f);
-                    AddWaypoint(2, 775.67f, 641.91f, 411.91f);
-                    uiWaypointPath = 3;
-                    break;
+                if (!pGrandChampion->HasAura(SPELL_KNEEL))
+                    return false;
             }
-
-            if (uiType <= 3)
-                Start(false, true);
         }
+        return true;
+    }
 
-        void WaypointReached(uint32 waypointId, uint32 /*pathId*/) override
+    void AttackRandomPlayer()
+    {
+        Map::PlayerList const &players = me->GetMap()->GetPlayers();
+        if (players.isEmpty())
+            return;
+
+        for (Map::PlayerList::const_iterator itr = players.begin(); itr != players.end(); ++itr)
         {
-            switch (waypointId)
+            Player *plr = itr->GetSource();
+            if (plr && !plr->IsGameMaster() && plr->IsAlive())
             {
-                case 2:
-                    if (uiWaypointPath == 3 || uiWaypointPath == 2)
-                        instance->SetData(DATA_MOVEMENT_DONE, instance->GetData(DATA_MOVEMENT_DONE)+1);
-                    break;
-                case 3:
-                    instance->SetData(DATA_MOVEMENT_DONE, instance->GetData(DATA_MOVEMENT_DONE)+1);
-                    break;
+                for (int i = 0; i < 3; i++)
+                {
+                    if (Creature* pGrandChampion = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_GRAND_CHAMPION_1 + i)))
+                    {
+                        EnterAggressiveMode(pGrandChampion);
+                        if (!pGrandChampion->IsInCombat())
+                            pGrandChampion->AI()->AttackStart(plr);
+                    }
+                }
+                break;
             }
         }
+    }
+
+    bool CanUseNormalAI()
+    {
+        // must check for unattackable flag to prevent crash in certain situations
```

Comment:
this was commented at https://github.com/TrinityCore/TrinityCore/pull/21427#discussion_r169184672 but it wasn't resolved